# The Analysis on the Quantity Comments with Title between Asking and Showing in Hacker News Community

Our aim in this project is to find the styles of title in the Hacker News Community (HNC) that have a great quantity in comments. In the HNC, there are many posts are being posted in the purpose of attracting the fellow heckers to comment or answer their questions.

In the community, it is important for us to know what types of submissions are popular. This means that, throughout this project, the finding will focus on the average of the comments for different types of submission and the time when the commented are posted

# The Opening and Exploring the Data

[A data set](https://www.kaggle.com/hacker-news/hacker-news-posts) is taken from a site, Hacker News which is a site started by the startup incubator [Y Combinator](https://www.ycombinator.com/). The Hacker News is knows as where user-submitted stories (known as "post") are voted and commented upon, similar to reddit. It is extremely popular in technology and startup circles, and posts that make is to the top of Hacker News' listings can get hundreds of thousands of visotors as a result.

Originally, There are almost 300,000 rows. However, due to some submissions that do not receive any comments, it is reduced to approximately 20,000 rows. Throught the database, we classify the submissions into two types, which are Asking and Showing. We would want to know which type of the submission received more comments on average, as well as the time when the comments are received.

Let's start with opening the data set, and  then contiune with exploring the dataset

In [1]:
from csv import reader

### The Hacker News Data set###
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
hn[:5] # open the first rows of the datase


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

We store the title of the dataset to `headers` and keep on the data in the `hn`

In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Next, we assort the data into `ask_post`, `show_posts`, and `other_posts`.

# Extracting Ask HN and Show HN Posts

First, we will identity the posts that begin either `Ask HN` or `Show HN` and saparate those datas into two different lists.

To do so, we will do as below

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


# Calculating the Average Numbers of comments for Ask HN and Show HN Posts

Now that we sepaprate ask posts and show posts, we will find the average number of comments for each type of posts

In [4]:
total_ask_comments = 0

for row in ask_posts:
    total_ask_comments += int(row[4])

avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [5]:
total_show_comments = 0

for row in show_posts:
    total_show_comments += int(row[4])

avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


On average, the `Ask HN` posts receive approximately 14.04 comments, where the `Show HN` posts receive about 10.32 comments. Therefore, the `Ask HN` posts receive more comments than the `Show HN` posts. We'll continue analyze on these numbers

# Finding the Amount Ask Posts and Comment by Hours Created

Next, we'll determine if we can maximize the quantity of comments in the ask posts by creating it in a certain period of time. First, we calculate the amount of ask posts created in each hour of the day, along with the number of comments received. Then, we calculate the average number of comments ask posts receive by hour created.

In [9]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received. 

import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [10]:
#Calculate the average amount of comments `Ask HN` posts created a each hour of the day receive.

avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr]/counts_by_hour[hr]])
    
avg_by_hour

[['15', 38.5948275862069],
 ['12', 9.41095890410959],
 ['18', 13.20183486238532],
 ['09', 5.5777777777777775],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['00', 8.127272727272727],
 ['11', 11.051724137931034],
 ['08', 10.25],
 ['06', 9.022727272727273],
 ['23', 7.985294117647059],
 ['04', 7.170212765957447],
 ['17', 11.46],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['02', 23.810344827586206],
 ['21', 16.009174311926607],
 ['16', 16.796296296296298],
 ['13', 14.741176470588234],
 ['22', 6.746478873239437],
 ['05', 10.08695652173913],
 ['07', 7.852941176470588],
 ['03', 7.796296296296297],
 ['20', 21.525]]

# Sorting and Printing Values from a List of Lists

In [12]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse = True)

sorted_swap

[[38.5948275862069, '15'], [9.41095890410959, '12'], [13.20183486238532, '18'], [5.5777777777777775, '09'], [10.8, '19'], [11.383333333333333, '01'], [8.127272727272727, '00'], [11.051724137931034, '11'], [10.25, '08'], [9.022727272727273, '06'], [7.985294117647059, '23'], [7.170212765957447, '04'], [11.46, '17'], [13.440677966101696, '10'], [13.233644859813085, '14'], [23.810344827586206, '02'], [16.009174311926607, '21'], [16.796296296296298, '16'], [14.741176470588234, '13'], [6.746478873239437, '22'], [10.08695652173913, '05'], [7.852941176470588, '07'], [7.796296296296297, '03'], [21.525, '20']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [13]:
# Sort the value and print the 5 hours with the highest average comments.

print("Top 5 Hours for `Ask HN` Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
                dt.datetime.strptime(hr, "%H").strftime("%H:%M"), avg
        )
    )

Top 5 Hours for `Ask HN` Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that received the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion

In the project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that *of the posts that received comments*, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).